File: fixtures/scale/php/medium/src/Repl/Repl.php
<?php

namespace Calc\Repl;

use Calc\Eval\Evaluator;
use Calc\Output\ResultFormatter;
use Calc\Parsing\Parser;

require_once __DIR__ . '/../Parsing/Parser.php';
require_once __DIR__ . '/../Eval/Evaluator.php';
require_once __DIR__ . '/../Output/ResultFormatter.php';
require_once __DIR__ . '/CommandRouter.php';

class Repl
{
    private InputReader $reader;

    public function __construct(private ReplContext $context)
    {
        $this->reader = new InputReader('calc> ');
    }

    public function run(): void
    {
        $evaluator = new Evaluator($this->context->environment, $this->context->functions);
        while ($this->context->settings->running && ($line = $this->reader->nextLine()) !== false) {
            $line = trim($line);
            if ($line === '') { continue; }
            if (str_starts_with($line, ':')) {
                echo CommandRouter::dispatch($line, $this->context) . "\n";
                continue;
            }
            $value = $evaluator->eval((new Parser($line))->parseStatement());
            $this->context->history->append($line, $value);
            echo ResultFormatter::formatResult($value, $this->context->settings) . "\n";
        }
    }
}

File: fixtures/scale/csharp/medium/Repl/Repl.cs
using System;
using Calc.Eval;
using Calc.Output;
using Calc.Parsing;

namespace Calc.Repl
{
    public class Repl
    {
        private readonly ReplContext _context;
        private readonly InputReader _reader;

        public Repl(ReplContext context)
        {
            _context = context;
            _reader = new InputReader("calc> ");
        }

        public void Run()
        {
            var evaluator = new Evaluator(_context.Environment, _context.Functions);
            string line;
            while (_context.Settings.Running && (line = _reader.NextLine()) != null)
            {
                if (line.Trim().Length == 0) continue;
                if (line.StartsWith(":"))
                {
                    Console.WriteLine(CommandRouter.Dispatch(line, _context));
                    continue;
                }
                var node = new Parser(line).ParseStatement();
                var value = evaluator.Eval(node);
                _context.History.Append(line.Trim(), value);
                Console.WriteLine(ResultFormatter.FormatResult(value, _context.Settings));
            }
        }
    }
}
